### PR TITLE
added payment gateway stripe 

### LINKED ISSUE #211: Add a payment gateway integration
**Describe the solution you'd like**
Implement a payment gateway system that will allow the organization to receive donations from users of the app

### PR DESCRIPTION
**Pull request by yasharth291**
What kind of change does this PR introduce?
It is an Enhancement

Summary

Adding payment gateway to take the donation from the user motivated me to raise this PR.

Other information
Fixes #211

Use this link to get to know about the stripe.js https://stripe.com/docs/js/

Changes I have done

Added the Stripe.js Payment gateway.
Inserted the Webhook for the notification and checking the endpoint
The webhooks can be saved in the mongo db or a mail for the same can be sent to the admin
Added the middleware to check the response before the payment is made so the user do not use it for spamming
Adding 3D checking model and SCA security
Adding a subscription model for the recurring payments
Creating the invoices and sending them to user of recurring payments

### CODE REVIEW

**Comment by yasharth291** [comment]
@Laurell876  sir i have rebased

**Comment by palisadoes** [comment]
@DMills27 @Laurell876 please review

**Comment by yasharth291** [comment]
@palisadoes  sir this is gsoc parallel issue

**Comment by yasharth291** [comment]
@palisadoes  sir could you please review it sir

File: resolvers/Payment_Gateway/checkout.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
+const stripe = require('stripe')('PUT YOUR SECURITY KEY HERE');
+const express = require("express");
+const router = express.Router();
+const { resolve } = require('path');
+router.post("/payment", 
+	[
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:6]: Pull the schema validation in some reusable middleware function

File: resolvers/Payment_Gateway/checkout.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		check('amount','amount is not in the body').not().isEmpty(),
+		check('id','The method of payment is not present').not().isEmpty()
+	],
+	async (req, res) => {
+		let {amount, id} = req.body //here amount is the amount for donation, and id is the way the user is going to pay like card UPI and other methods
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:11]: use const as you are not modifying the variables

File: resolvers/Payment_Gateway/checkout.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		} catch (error) {
+			console.log("Error", error)
+			res.json({
+				message: "Payment failed",
+				success: false
+			})
+		}
+	});
+	// Copy the .env.example in the root into a .env file in this folder
+	require('dotenv').config({ path: './.env' });
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:34]: checking env variables available or not is not required

File: resolvers/Payment_Gateway/checkout.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	
+	// Ensure environment variables are set.
+	checkEnv();
+	
+	const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:39]: const { STRIPE_SECRET_KEY } = require('config');

Comment (impiyush83) [reply resolvers/Payment_Gateway/checkout.js:39]: Config file must have all mappings of env vars: 

module.exports = { 
    STRIPE_SECRET_KEY: process.env.STRIPE_SECRET_KEY
};

File: resolvers/Payment_Gateway/checkout.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	
+	// Fetch the Checkout Session to display the JSON result on the success page
+	app.get('/checkout-session', async (req, res) => {
+	  const { sessionId } = req.query;
+	  const session = await stripe.checkout.sessions.retrieve(sessionId);
+	  res.send(session);
+	});
+	
+	app.post('/create-checkout-session', async (req, res) => {
+	  const domainURL = process.env.DOMAIN;
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:77]: pull DOMAIN out in config file as mentioned above

File: resolvers/Payment_Gateway/checkout.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	
+	  const { quantity, locale } = req.body;
+	
+	  // The list of supported payment method types. We fetch this from the
+	  // environment variables in this sample. In practice, users often hard code a
+	  // list of strings for the payment method types they plan to support.
+	  const pmTypes = (process.env.PAYMENT_METHOD_TYPES || 'card').split(',').map((m) => m.trim());
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:84]: pull this env variable in config file

File: resolvers/Payment_Gateway/checkout.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	
+	  // Create new Checkout Session for the order
+	  // Other optional params include:
+	  // [billing_address_collection] - to display billing address details on the page
+	  // [customer] - if you have an existing Stripe Customer ID
+	  // [customer_email] - lets you prefill the email input in the Checkout page
+	  // For full details see https://stripe.com/docs/api/checkout/sessions/create
+	  const session = await stripe.checkout.sessions.create({
+		payment_method_types: pmTypes,
+		mode: 'payment',
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:94]: payment is a constant pull that in constants file

File: resolvers/Payment_Gateway/checkout.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	});
+	
+	// Webhook handler for asynchronous events.
+	app.post('/webhook', async (req, res) => {
+	  let data;
+	  let eventType;
+	  // Check if webhook signing is configured.
+	  if (process.env.STRIPE_WEBHOOK_SECRET) {
+		// Retrieve the event by verifying the signature using the raw body and secret.
+		let event;
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/checkout.js:119]: use const whereever possible.

File: resolvers/Payment_Gateway/subscription.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      url: portalsession.url,
+    });
+  });
+  
+  // Webhook handler for asynchronous events.
+  app.post("/webhook", async (req, res) => {
+    let data;
+    let eventType;
+    // Check if webhook signing is configured.
+    if (process.env.STRIPE_WEBHOOK_SECRET) {
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/subscription.js:118]: pull this out in a separate function which spits out data and eventType,  at last assign those variables to data and eventType declaring as const

File: resolvers/Payment_Gateway/subscription.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+      data = event.data;
+      eventType = event.type;
+    } else {
+      // Webhook signing is recommended, but if the secret is not configured in `config.js`,
+      // retrieve the event data directly from the request body.
+      data = req.body.data;
+      eventType = req.body.type;
+    }
+  
+    if (eventType === "checkout.session.completed") {
```
Comment (impiyush83) [inline resolvers/Payment_Gateway/subscription.js:143]: pull out checkout.session.completed in constants folder -> eventTypes.js file

**Review by impiyush83 — CHANGES_REQUESTED** [review]
@yasharth291  Have added comments for code cleanup

**Pull request closed without merging**